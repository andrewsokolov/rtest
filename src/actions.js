export const UPDATE_EXCHANGE_RATES = 'UPDATE_EXCHANGE_RATES'
export const FETCH_EXCHANGE_RATES_START = 'FETCH_EXCHANGE_RATES_START'
export const FETCH_EXCHANGE_RATES_SUCCESS = 'FETCH_EXCHANGE_RATES_SUCCESS'
export const FETCH_EXCHANGE_RATES_FAILURE = 'FETCH_EXCHANGE_RATES_FAILURE'
export const EXCHANGE = 'EXCHANGE'
export const SWITCH_SECTIONS = 'SWITCH_SECTIONS'
export const SET_INPUT_VALUE = 'SET_INPUT_VALUE'
export const SET_ACTIVE_SECTION = 'SET_ACTIVE_SECTION'
export const SELECT_CURRENCY = 'SELECT_CURRENCY'

export const TYPES = {
    TARGET: "target",
    SOURCE: "source"
}

export function updateExchangeRates(data) {
    return { type: UPDATE_EXCHANGE_RATES, data }
}

export function exchange() {
    return (dispatch, getState) => {
        const sourceCurrency = getState().currencies.source
        const targetCurrency = getState().currencies.target
        const targetExchangeRate = getState().exchange.rates[targetCurrency]
        const sourceValue = getState().values.source
        // const targetValue = getState().values.target
        const sourceBalance = getState().balance[sourceCurrency]
        const targetBalance = getState().balance[targetCurrency]
        if (sourceValue > sourceBalance) {
            return
        }
        const newSourceBalance = sourceBalance - sourceValue
        const newTargetBalance = targetBalance + (sourceValue * targetExchangeRate)
        dispatch({ type: EXCHANGE, payload: { [sourceCurrency]: newSourceBalance, [targetCurrency]: newTargetBalance } })
        dispatch(changeInputValue(0, TYPES.SOURCE))
    }
}

export function setInputValue(value, section) {
    return { type: SET_INPUT_VALUE, payload: { value, section } }
}

export function setActiveSection(section) {
    return { type: SET_ACTIVE_SECTION, payload: { section } }
}

export function selectCurrency(currency, section) {
    return { type: SELECT_CURRENCY, payload: { currency, section } }
}

export function changeInputValue(value, section) {
    return (dispatch, getState) => {
        dispatch(setActiveSection(section))
        const oppositeSection = section === TYPES.TARGET ? TYPES.SOURCE : TYPES.TARGET
        dispatch(setInputValue(value, section))
        const targetCurrency = getState().currencies.target
        const exchangeRate = getState().exchange.rates[targetCurrency]
        const newValue = section === TYPES.TARGET ? value / exchangeRate : value * exchangeRate
        dispatch(setInputValue(Number(newValue.toFixed(2)), oppositeSection))
    }
}

export function setEditableValue(section) {
    return (dispatch, getState) => {
        const sectionValue = getState().values[section]
        dispatch(changeInputValue(sectionValue, section))
    }
}


export function switchSections() {
    return (dispatch, getState) => {
        dispatch({ type: SWITCH_SECTIONS })
        dispatch(fetchExchangeRates(getState().currencies.source)).then((data) => {
            const activeSection = getState().activeSection
            const sectionValue = getState().values[activeSection]
            dispatch(changeInputValue(sectionValue, activeSection))
        })
    }
}

export function fetchExchangeRates(base) {
    return (dispatch, getState) => {
        if (isBaseSame(getState(), base) && isLoading(getState())) {
            return
        }
        dispatch({ type: FETCH_EXCHANGE_RATES_START, payload: { base } })
        return fetch(`https://api.exchangeratesapi.io/latest?base=${base}`)
            .then((response) => response.json())
            .then(({ rates }) => {
                if (!isBaseSame(getState(), base)) {
                    return
                }
                dispatch({ type: FETCH_EXCHANGE_RATES_SUCCESS, payload: { base, rates } })
            })
            .catch((error) => {
                if (!isBaseSame(getState(), base)) {
                    return
                }
                dispatch({ type: FETCH_EXCHANGE_RATES_FAILURE, payload: { base, error } })
            })
    }
}



export function isBaseSame(state, base) {
    return state.exchange.base === base
}

export function isLoading(state) {
    return state.exchange.isLoading
}