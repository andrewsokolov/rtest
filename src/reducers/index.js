import { FETCH_EXCHANGE_RATES_START, FETCH_EXCHANGE_RATES_SUCCESS, FETCH_EXCHANGE_RATES_FAILURE, SET_ACTIVE_SECTION, EXCHANGE, SWITCH_SECTIONS, SET_INPUT_VALUE, SELECT_CURRENCY, TYPES } from "../actions"
import { combineReducers } from 'redux'


const defaultBalance = {
  "GBP": 100,
  "USD": 0,
  "EUR": 0
}

function activeSection(state = TYPES.SOURCE, action) {
  switch (action.type) {
    case SET_ACTIVE_SECTION:
      return action.payload.section
    default:
      return state
  }
}

function balance(state = defaultBalance, action) {
  switch (action.type) {
    case EXCHANGE:
      return { ...state, ...action.payload }
    default:
      return state
  }
}

function exchange(state = { rates: null, base: null, error: null, isLoading: false }, action) {
  switch (action.type) {
    case FETCH_EXCHANGE_RATES_START:
      return {
        ...state,
        base: action.payload.base,
        error: null,
        isLoading: true
      }
    case FETCH_EXCHANGE_RATES_SUCCESS:
      return {
        base: action.payload.base,
        rates: action.payload.rates,
        isLoading: false
      }
    case FETCH_EXCHANGE_RATES_FAILURE:
      return {
        ...state,
        base: action.payload.base,
        error: action.payload.error,
        isLoading: false
      }
    default:
      return state
  }
}

function values(state = { target: 0, source: 0 }, action) {
  switch (action.type) {
    case SWITCH_SECTIONS:
      return {
        [TYPES.SOURCE]: state.target,
        [TYPES.TARGET]: state.source
      }
    case SET_INPUT_VALUE:
      return {
        ...state,
        [action.payload.section]: action.payload.value
      }
    default:
      return state
  }
}

function currencies(state = { target: "GBP", source: "EUR" }, action) {
  switch (action.type) {
    case SWITCH_SECTIONS:
      const newSource = state.target
      return {
        source: newSource,
        target: state.source
      }
    case SELECT_CURRENCY:
      return {
        ...state,
        [action.payload.section]: action.payload.currency
      }
    default:
      return state
  }
}

const app = combineReducers({
  balance,
  exchange,
  values,
  currencies,
  activeSection
})

export { app }