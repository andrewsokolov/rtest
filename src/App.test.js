import { updateExchangeRates, setEditableValue, changeInputValue, TYPES, setInputValue, switchAccounts, selectCurrency, switchSections } from './actions'
import { createStore, applyMiddleware } from 'redux'
import { app } from './reducers'
import ReduxThunk from 'redux-thunk';

const rates = {
  base: "EUR",
  rates: {
    "GBP": 1.2,
    "USD": 1.1,
    "EUR": 1
  }
}

test('renders learn react link', () => {
  const store = createStore(app, applyMiddleware(ReduxThunk))
  store.dispatch(updateExchangeRates(rates))
  expect(store.getState()).toMatchSnapshot()
  store.dispatch(setInputValue(20, TYPES.SOURCE))
  expect(store.getState()).toMatchSnapshot()
  store.dispatch(setInputValue(20, TYPES.TARGET))
  expect(store.getState()).toMatchSnapshot()
  store.dispatch(selectCurrency("USD", TYPES.TARGET))
  expect(store.getState()).toMatchSnapshot()
  store.dispatch(selectCurrency("EUR", TYPES.SOURCE))
  expect(store.getState()).toMatchSnapshot()
  store.dispatch(changeInputValue(100, TYPES.SOURCE))
  expect(store.getState()).toMatchSnapshot()
  store.dispatch(changeInputValue(200, TYPES.TARGET))
  expect(store.getState()).toMatchSnapshot()
  store.dispatch(setEditableValue(TYPES.SOURCE))
  expect(store.getState()).toMatchSnapshot()
  store.dispatch(setEditableValue(TYPES.TARGET))
  expect(store.getState()).toMatchSnapshot()


  store.dispatch(switchSections())
  expect(store.getState()).toMatchSnapshot()
  store.dispatch(switchSections())
  expect(store.getState()).toMatchSnapshot()
});
