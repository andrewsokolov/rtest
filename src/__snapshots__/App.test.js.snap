// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders learn react link 1`] = `
Object {
  "activeSection": "source",
  "balance": Object {
    "EUR": 0,
    "GBP": 100,
    "USD": 0,
  },
  "currencies": Object {
    "source": "EUR",
    "target": "GBP",
  },
  "exchange": Object {
    "base": "EUR",
    "rates": Object {
      "EUR": 1,
      "GBP": 1.2,
      "USD": 1.1,
    },
  },
  "values": Object {
    "source": 0,
    "target": 0,
  },
}
`;

exports[`renders learn react link 2`] = `
Object {
  "activeSection": "source",
  "balance": Object {
    "EUR": 0,
    "GBP": 100,
    "USD": 0,
  },
  "currencies": Object {
    "source": "EUR",
    "target": "GBP",
  },
  "exchange": Object {
    "base": "EUR",
    "rates": Object {
      "EUR": 1,
      "GBP": 1.2,
      "USD": 1.1,
    },
  },
  "values": Object {
    "source": 20,
    "target": 0,
  },
}
`;

exports[`renders learn react link 3`] = `
Object {
  "activeSection": "source",
  "balance": Object {
    "EUR": 0,
    "GBP": 100,
    "USD": 0,
  },
  "currencies": Object {
    "source": "EUR",
    "target": "GBP",
  },
  "exchange": Object {
    "base": "EUR",
    "rates": Object {
      "EUR": 1,
      "GBP": 1.2,
      "USD": 1.1,
    },
  },
  "values": Object {
    "source": 20,
    "target": 20,
  },
}
`;

exports[`renders learn react link 4`] = `
Object {
  "activeSection": "source",
  "balance": Object {
    "EUR": 0,
    "GBP": 100,
    "USD": 0,
  },
  "currencies": Object {
    "source": "EUR",
    "target": "USD",
  },
  "exchange": Object {
    "base": "EUR",
    "rates": Object {
      "EUR": 1,
      "GBP": 1.2,
      "USD": 1.1,
    },
  },
  "values": Object {
    "source": 20,
    "target": 20,
  },
}
`;

exports[`renders learn react link 5`] = `
Object {
  "activeSection": "source",
  "balance": Object {
    "EUR": 0,
    "GBP": 100,
    "USD": 0,
  },
  "currencies": Object {
    "source": "EUR",
    "target": "USD",
  },
  "exchange": Object {
    "base": "EUR",
    "rates": Object {
      "EUR": 1,
      "GBP": 1.2,
      "USD": 1.1,
    },
  },
  "values": Object {
    "source": 20,
    "target": 20,
  },
}
`;

exports[`renders learn react link 6`] = `
Object {
  "activeSection": "source",
  "balance": Object {
    "EUR": 0,
    "GBP": 100,
    "USD": 0,
  },
  "currencies": Object {
    "source": "EUR",
    "target": "USD",
  },
  "exchange": Object {
    "base": "EUR",
    "rates": Object {
      "EUR": 1,
      "GBP": 1.2,
      "USD": 1.1,
    },
  },
  "values": Object {
    "source": 100,
    "target": "110.00",
  },
}
`;

exports[`renders learn react link 7`] = `
Object {
  "activeSection": "target",
  "balance": Object {
    "EUR": 0,
    "GBP": 100,
    "USD": 0,
  },
  "currencies": Object {
    "source": "EUR",
    "target": "USD",
  },
  "exchange": Object {
    "base": "EUR",
    "rates": Object {
      "EUR": 1,
      "GBP": 1.2,
      "USD": 1.1,
    },
  },
  "values": Object {
    "source": "181.82",
    "target": 200,
  },
}
`;

exports[`renders learn react link 8`] = `
Object {
  "activeSection": "source",
  "balance": Object {
    "EUR": 0,
    "GBP": 100,
    "USD": 0,
  },
  "currencies": Object {
    "source": "EUR",
    "target": "USD",
  },
  "exchange": Object {
    "base": "EUR",
    "rates": Object {
      "EUR": 1,
      "GBP": 1.2,
      "USD": 1.1,
    },
  },
  "values": Object {
    "source": "181.82",
    "target": "200.00",
  },
}
`;

exports[`renders learn react link 9`] = `
Object {
  "activeSection": "target",
  "balance": Object {
    "EUR": 0,
    "GBP": 100,
    "USD": 0,
  },
  "currencies": Object {
    "source": "EUR",
    "target": "USD",
  },
  "exchange": Object {
    "base": "EUR",
    "rates": Object {
      "EUR": 1,
      "GBP": 1.2,
      "USD": 1.1,
    },
  },
  "values": Object {
    "source": "181.82",
    "target": "200.00",
  },
}
`;

exports[`renders learn react link 10`] = `
Object {
  "activeSection": "source",
  "balance": Object {
    "EUR": 0,
    "GBP": 100,
    "USD": 0,
  },
  "currencies": Object {
    "source": "USD",
    "target": "EUR",
  },
  "exchange": Object {
    "base": "EUR",
    "rates": Object {
      "EUR": 1,
      "GBP": 1.2,
      "USD": 1.1,
    },
  },
  "values": Object {
    "source": "200.00",
    "target": "200.00",
  },
}
`;

exports[`renders learn react link 11`] = `
Object {
  "activeSection": "target",
  "balance": Object {
    "EUR": 0,
    "GBP": 100,
    "USD": 0,
  },
  "currencies": Object {
    "source": "EUR",
    "target": "USD",
  },
  "exchange": Object {
    "base": "EUR",
    "rates": Object {
      "EUR": 1,
      "GBP": 1.2,
      "USD": 1.1,
    },
  },
  "values": Object {
    "source": "181.82",
    "target": "200.00",
  },
}
`;
